reportportal_client/portal.py

```diff
@@ -0,0 +1,82 @@
+            endpoint=endpoint, project=project, token=token, verify_ssl=verify_ssl, **kwargs)
+
+    @retry(
+        stop=stop_after_attempt(3),
+        wait=wait_fixed(5),
+    )
+    def get_launches(self, page=1, page_size=50, page_sort='startTime,desc', **filters):
+        """Returns launches in Report Portal customized by sat_version, launch_type and
+        latest number of launches sorted by latest sat version/snap version.
+        This does not includes each tests data for all tests in all launches, but it includes
+        just an overview data for all and each launch in Report Portal
+        :param str sat_version: The satellite version
+            If its not specified, then latest count of launches of satellite versions returned
+        :param str launch_type: Either satellite6 or upgrades,
+            default returns only non-upgrade launches
+        :returns dict: The launches of Report portal.
+            if sat_version is given,
+            ```{'snap_version1':launch_object1, 'snap_version2':launch_object2}```
+            else,
+            ```{'sat_version1':{'snap_version1':launch_object1, ..}, 'sat_version2':{}}```
+        """
+        params = {'page.page': page, 'page.size': page_size, 'page.sort': page_sort}
+        params.update(filters)
+        launches = []
+        total_pages = page
+        while page <= total_pages:
+            params['page.page'] = page
+            resp = self.read(component="launch", **params)
+            attr_launches = [launch for launch in resp.json()['content']]
+            launches.extend([AttrDict(lnch) for lnch in attr_launches])
+            total_pages = resp.json()['page']['totalPages']
+            page += 1
+        return launches
+
+    @retry(
+        stop=stop_after_attempt(3),
+        wait=wait_fixed(5),
+    )
+    def get_tests(self, launch_id, isLatest=False, page=1, page_size=100, page_sort='desc', **filters):
+        """Returns tests data customized by kwargs parameters.
+        This is a main function that will be called to retrieve the tests data
+        of a particular test status or/and defect_type
+        :param str status: Filter tests of a launch with tests `status`
+        :param str defect_type: Filter tests of a launch with tests `defect_type`
+        :returns dict: All filtered tests dict based on params data keyed by test name and test
+            properties as value, in format -
+            ```{'test_name1':test1_properties_dict, 'test_name2':test2_properties_dict}```
+        """
+        params = {
+            'filter.eq.launchId': launch_id, 'isLatest': isLatest, 'page.page': page,
+            'page.size': page_size, 'page.sort': page_sort}
+        params.update(filters)
+        tests = []
+        total_pages = page
+        while page <= total_pages:
+            params['page.page'] = page
+            resp = self.read(component='item', **params)
+            tests.extend([AttrDict(tst) for tst in resp.json()['content']])
```

Comment:
Why is it necessary to pack dictionaries into AttrDict objects?

Reply:
This is nothing but convenience to use the dictionary keys as attributes instead of plain dictionary keys.

`test.id` instead of `test['id']`.